capi_problems.rst

```diff
@@ -70,7 +70,141 @@ and the particular requirements that each of them has.
+actions that different stakeholders need to perform through
+the C API.
+
+There are actions which are generic, and required by
+all types of API users:
+
+* Define functions and call them
+* Create classes and instantiate them
+* Create instances of builtin and user defined types
+  and perform operations on them
+* Introspect objects, including types, instances, and functions.
+* Raise and handle exceptions
+* Import modules
+* Manage threads
+* Manage sub-interpreters
+* Handle and send signals
+
+Groups of users are:
+
+**Extension writers**
+
+These are the traditional users of the C API, and their requirements
+are listed above.
+
+**Embedders**
+
+Applications with an embedded Python interpreter. Example are
+`Blender <https://docs.blender.org/api/current/info_overview.html>`__ and
+`OBS <https://obsproject.com/wiki/Getting-Started-With-OBS-Scripting>`__.
+
+They need to be able to
+
+* Configure the interpreter (import paths, inittab, sys.argv, ...)
+* Interact with the execution model and program lifetime, including
+  clean interpreter shutdown and restart
+* Represent complex data models in a way Python can use without
+  having to create deep copies.
+* Frozen modules.
+* Run multiple independent interpreters (in particular, when embedded
+  in a library that wants to avoid global effects).
+
+**Alternative Python Implementations**
+
+Alterantive implementations of Python (such as
+`PyPy <https://www.pypy.org>`__,
+`GraalPython <https://www.graalvm.org/python/>`__,
+and `Jython <https://www.jython.org>`__), may take
+very different approaches for the implementation of
+different subsystems. They need:
+
+* The API to be abstract and hide implementation details.
+* A specification of the API, ideally with a test suite
+  that ensures compatibility.
+* An ABI that can be shared across Python implementations.
+
+**Alternative APIs**
+
+There are several projects that implement alternatives to the
+C API, which offer extension users advantanges over programming
+directly with the C API. These APIs are implemented with the
+C API, and in some cases by using cpython internals.
+Some examples are
+`Cython <https://cython.org>`__,
+`HPy <https://hpyproject.org>`__,
+`pythoncapi-compat <https://pythoncapi-compat.readthedocs.io/en/latest/>`__
+and CPython's own
+`Argument Clinic <https://docs.python.org/3/howto/clinic.html>`__.
+
+Such systems need minimal building blocks for accessing CPython
+efficiently. They don't necessarily need an ergonomic API, because
+they typically generated code that is not intended to be read
+by humans. But they do need it to be comprehesive enough so that
+they don't need to access internals, while offering them stability,
+and without sacrificing performance.
+
+An alternative is to have a fast API tier with lower stability
+guarantees. Then the users of these tools can choose whether
+to generate code that uses the faster or the more stable version
+of the API.
```

Comment:
I saw on on going discussion about "unsafe" API which may or not be faster. I don't think that the main difference is about backward compatibilty. It's more about checking pre-conditions and post-conditions or not. Usually, unsafe functions require the caller to avoid calling such functions unless multiple conditions are respected. If you respect these conditions, the API is safe and may stay backward compatible for a long time (ex: PyTuple_SET_ITEM id there for 30 years and didn't change).

Or are you referring to the unstable API which is more about exposing internals which are known to evolve frequently like AST?

Reply:
> Or are you referring to the unstable API which is more about exposing internals which are known to evolve frequently like AST?


Both. I will reword this paragraph.